File: C/3_sendReceive/send_receive.c
#include <stdio.h>
#include <stdlib.h>

#include <mpi.h>

int main(int argc, char **argv)
{
    int maxElements = 1048576;

    int *data = (int *)malloc(sizeof(int) * maxElements);
    if (data == NULL)
    {
        printf("Not enough memory\n");
        return -1;
    }
    else
    {
        // Initialize with random data for this example
        srand(42);
        for (int i = 0; i < maxElements; ++i)
        {
            data[i] = rand();
        }
    }

    int worldRank, worldSize;
    // MPI Initialization
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &worldSize);
    MPI_Comm_rank(MPI_COMM_WORLD, &worldRank);

    if (worldSize != 2)
    {
        printf("This program can only be started with 2 MPI processes\n");
        MPI_Abort(MPI_COMM_WORLD, 1);
        return 1;
    }

    int nextRank = (worldRank + 1) % worldSize;
    int prevRank = (worldRank - 1 + worldSize) % worldSize;
    int numElements;

    // TODO: Remove the deadlock
    for (numElements = 4096; numElements < maxElements; numElements += 4096)
    {
        MPI_Status status;
        MPI_Request request;

        printf("Rank %i sends %i elements of data now\n",
            worldRank, numElements);
        MPI_Isend(data, numElements, MPI_INT, nextRank, 0, MPI_COMM_WORLD, &request);

        printf("Rank %i receives %i elements of data now\n",
            worldRank, numElements);
        MPI_Irecv(data, numElements, MPI_INT, prevRank, 0,
            MPI_COMM_WORLD, &request);

        printf("Rank %i is done with %i elements of data\n",
            worldRank, numElements);
        
        MPI_Wait(&request, &status);
    }

    MPI_Finalize();
    return 0;
}

File: Fortran/3_sendReceive/send_receive.f90
program sendReceive
    use mpi
    implicit none

    integer, parameter :: maxElements = 1048576

    integer :: myRank, numProcs, ierr
    integer, allocatable, dimension(:) :: data
    integer :: nextRank, prevRank, numElements
    integer, dimension(MPI_STATUS_SIZE) :: status

    allocate(data(maxElements))
    if (.not.allocated(data)) then
        print *,"Not enough memory"
        stop 1
    end if
    ! Note: Data remains uninitialized, as this is just an example

    ! MPI Initialization
    call MPI_INIT(ierr)
    call MPI_COMM_RANK(MPI_COMM_WORLD, myRank, ierr)
    call MPI_COMM_SIZE(MPI_COMM_WORLD, numProcs, ierr)

    if (numProcs /= 2) then
        print *,"Error: This program can only run with 2 MPI tasks!"
        call MPI_ABORT(MPI_COMM_WORLD, 1, ierr)
        stop 1
    end if

    nextRank = mod(myRank+1, numProcs)
    prevRank = mod(myRank + numProcs - 1, numProcs)

    ! TODO: Remove the deadlock
    do numElements = 4096, maxElements, 4096
        print '("Process ",I0," sends ",I0," elements of data now")', &
            myRank, numElements
        call MPI_SEND(data, numElements, MPI_INTEGER, nextRank, 0, &
            MPI_COMM_WORLD, ierr)
        print '("Process ",I0," receives ",I0," elements of data now")', &
            myRank, numElements
        call MPI_RECV(data, numElements, MPI_INTEGER, prevRank, 0, &
            MPI_COMM_WORLD, status, ierr)
        print '("Process ",I0," is done with ",I0," elements of data")', &
            myRank, numElements
    end do

    deallocate(data)

    call MPI_FINALIZE(ierr)
end program sendReceive
